demos/device_defender_for_aws/defender_demo.c

```diff
@@ -0,0 +1,697 @@
+    DefenderStatus_t status;
+    DefenderTopic_t api;
+    bool validationResult;
+    MQTTPublishInfo_t * pPublishInfo = pDeserializedInfo->pPublishInfo;
+
+    /* Silence compiler warnings about unused variables. */
+    ( void ) pMqttContext;
+
+    /* Handle incoming publish. The lower 4 bits of the publish packet
+     * type is used for the dup, QoS, and retain flags. Hence masking
+     * out the lower bits to check if the packet is publish. */
+    if( ( pPacketInfo->type & 0xF0U ) == MQTT_PACKET_TYPE_PUBLISH )
+    {
+        status = Defender_MatchTopic( pPublishInfo->pTopicName,
+                                      pPublishInfo->topicNameLength,
+                                      &( api ),
+                                      NULL,
+                                      NULL );
+
+        if( status == DefenderSuccess )
+        {
+            if( api == DefenderJsonReportAccepted )
+            {
+                /* Check if the response is valid and is for the report we published. */
+                validationResult = validateDefenderResponse( pPublishInfo->pPayload,
+                                                             pPublishInfo->payloadLength );
+
+                if( validationResult == true )
+                {
+                    LogInfo( ( "The defender report was accepted by the service. Response: %.*s.",
+                               ( int ) pPublishInfo->payloadLength,
+                               ( const char * ) pPublishInfo->pPayload ) );
+                    reportStatus = ReportStatusAccepted;
+                }
+            }
+            else if( api == DefenderJsonReportRejected )
+            {
+                /* Check if the response is valid and is for the report we published. */
+                validationResult = validateDefenderResponse( pPublishInfo->pPayload,
+                                                             pPublishInfo->payloadLength );
+
+                if( validationResult == true )
+                {
+                    LogError( ( "The defender report was rejected by the service. Response: %.*s.",
+                                ( int ) pPublishInfo->payloadLength,
+                                ( const char * ) pPublishInfo->pPayload ) );
+                    reportStatus = ReportStatusRejected;
+                }
+            }
+            else
+            {
+                LogError( ( "Unexpected defender API : %d.", api ) );
+            }
+        }
+        else
+        {
+            LogError( ( "Unexpected publish message received. Topic: %.*s, Payload: %.*s.",
+                        ( int ) pPublishInfo->topicNameLength,
+                        ( const char * ) pPublishInfo->pTopicName,
+                        ( int ) pPublishInfo->payloadLength,
+                        ( const char * ) ( pPublishInfo->pPayload ) ) );
+        }
+    }
+    else
+    {
+        vHandleOtherIncomingPacket( pPacketInfo, pDeserializedInfo->packetIdentifier );
```

Comment:
Where is this function defined? I couldn't find it

Reply:
demos/common/mqtt_demo_helpers/mqtt_demo_helpers.h